In [ ]:
# 1: Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set display options
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

print("Libraries imported successfully")

Libraries imported successfully


In [2]:
# 2: Load the data
df = pd.read_csv('../data/raw/insurance_claims_raw.csv')

print(f"Dataset loaded successfully")
print(f"Shape: {df.shape}")
print(f"Rows: {df.shape[0]:,}")
print(f"Columns: {df.shape[1]}")

Dataset loaded successfully
Shape: (15420, 33)
Rows: 15,420
Columns: 33


In [3]:
# 3: First look at the data
print("First 5 rows:")
df.head()

First 5 rows:


Month  WeekOfMonth  DayOfWeek    Make AccidentArea DayOfWeekClaimed  \
0   Dec            5  Wednesday   Honda        Urban          Tuesday   
1   Jan            3  Wednesday   Honda        Urban           Monday   
2   Oct            5     Friday   Honda        Urban         Thursday   
3   Jun            2   Saturday  Toyota        Rural           Friday   
4   Jan            5     Monday   Honda        Urban          Tuesday   

  MonthClaimed  WeekOfMonthClaimed     Sex MaritalStatus  Age          Fault  \
0          Jan                   1  Female        Single   21  Policy Holder   
1          Jan                   4    Male        Single   34  Policy Holder   
2          Nov                   2    Male       Married   47  Policy Holder   
3          Jul                   1    Male       Married   65    Third Party   
4          Feb                   2  Female        Single   27    Third Party   

          PolicyType VehicleCategory      VehiclePrice  PolicyNumber  \
0  Sport - Liability           Sport  more than 69,000             1   
1  Sport - Collision           Sport  more than 69,000             2   
2  Sport - Collision           Sport  more than 69,000             3   
3  Sedan - Liability           Sport  20,000 to 29,000             4   
4  Sport - Collision           Sport  more than 69,000             5   

   RepNumber  Deductible  DriverRating Days:Policy-Accident Days:Policy-Claim  \
0         12         300             1         more than 30      more than 30   
1         15         400             4         more than 30      more than 30   
2          7         400             3         more than 30      more than 30   
3          4         400             2         more than 30      more than 30   
4          3         400             1         more than 30      more than 30   

  PastNumberOfClaims AgeOfVehicle AgeOfPolicyHolder PoliceReportFiled  \
0               none      3 years          26 to 30                No   
1               none      6 years          31 to 35               Yes   
2                  1      7 years          41 to 50                No   
3                  1  more than 7          51 to 65               Yes   
4               none      5 years          31 to 35                No   

  WitnessPresent AgentType NumberOfSuppliments AddressChange-Claim  \
0             No  External                none              1 year   
1             No  External                none           no change   
2             No  External                none           no change   
3             No  External         more than 5           no change   
4             No  External                none           no change   

  NumberOfCars  Year BasePolicy FraudFound  
0       3 to 4  1994  Liability         No  
1    1 vehicle  1994  Collision         No  
2    1 vehicle  1994  Collision         No  
3    1 vehicle  1994  Liability         No  
4    1 vehicle  1994  Collision         No

In [5]:
# 4: Data types and info
print("Dataset Info:")
df.info()

Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15420 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Month                 15420 non-null  object
 1   WeekOfMonth           15420 non-null  int64 
 2   DayOfWeek             15420 non-null  object
 3   Make                  15420 non-null  object
 4   AccidentArea          15420 non-null  object
 5   DayOfWeekClaimed      15420 non-null  object
 6   MonthClaimed          15420 non-null  object
 7   WeekOfMonthClaimed    15420 non-null  int64 
 8   Sex                   15420 non-null  object
 9   MaritalStatus         15420 non-null  object
 10  Age                   15420 non-null  int64 
 11  Fault                 15420 non-null  object
 12  PolicyType            15420 non-null  object
 13  VehicleCategory       15420 non-null  object
 14  VehiclePrice          15420 non-null  object
 15  PolicyNumber          

In [6]:
# 5: Column names
print("All columns:")
print(df.columns.tolist())

All columns:
['Month', 'WeekOfMonth', 'DayOfWeek', 'Make', 'AccidentArea', 'DayOfWeekClaimed', 'MonthClaimed', 'WeekOfMonthClaimed', 'Sex', 'MaritalStatus', 'Age', 'Fault', 'PolicyType', 'VehicleCategory', 'VehiclePrice', 'PolicyNumber', 'RepNumber', 'Deductible', 'DriverRating', 'Days:Policy-Accident', 'Days:Policy-Claim', 'PastNumberOfClaims', 'AgeOfVehicle', 'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType', 'NumberOfSuppliments', 'AddressChange-Claim', 'NumberOfCars', 'Year', 'BasePolicy', 'FraudFound']


In [ ]:
# 6: Check for target variable
# Look for fraud column (might be named: fraud, fraud_reported, fraudulent, etc.)-assuming we don't know
fraud_columns = [col for col in df.columns if 'fraud' in col.lower()]
print(f"Fraud-related columns: {fraud_columns}")

# Check the fraud distribution
if fraud_columns:
    fraud_col = fraud_columns[0]
    print(f"\nFraud distribution in '{fraud_col}':")
    print(df[fraud_col].value_counts())
    print(f"\nFraud percentage: {df[fraud_col].value_counts(normalize=True) * 100}")

Fraud-related columns: ['FraudFound']

Fraud distribution in 'FraudFound':
FraudFound
No     14497
Yes      923
Name: count, dtype: int64

Fraud percentage: FraudFound
No     94.014267
Yes     5.985733
Name: proportion, dtype: float64


In [ ]:
# 7: Missing values check
print("Missing values:")
missing = df.isnull().sum()
missing_pct = (missing / len(df)) * 100
missing_df = pd.DataFrame({
    'Missing Count': missing,
    'Percentage': missing_pct
})
missing_df = missing_df[missing_df['Missing Count'] > 0].sort_values('Missing Count', ascending=False)
print(missing_df)

Missing values:
Empty DataFrame
Columns: [Missing Count, Percentage]
Index: []


In [11]:
# 8: Basic statistics
print("Numerical columns statistics:")
df.describe()

Numerical columns statistics:


WeekOfMonth  WeekOfMonthClaimed           Age  PolicyNumber  \
count  15420.000000        15420.000000  15420.000000  15420.000000   
mean       2.788586            2.693969     39.855707   7710.500000   
std        1.287585            1.259115     13.492377   4451.514911   
min        1.000000            1.000000      0.000000      1.000000   
25%        2.000000            2.000000     31.000000   3855.750000   
50%        3.000000            3.000000     38.000000   7710.500000   
75%        4.000000            4.000000     48.000000  11565.250000   
max        5.000000            5.000000     80.000000  15420.000000   

          RepNumber    Deductible  DriverRating          Year  
count  15420.000000  15420.000000  15420.000000  15420.000000  
mean       8.483268    407.704280      2.487808   1994.866472  
std        4.599948     43.950998      1.119453      0.803313  
min        1.000000    300.000000      1.000000   1994.000000  
25%        5.000000    400.000000      1.000000   1994.000000  
50%        8.000000    400.000000      2.000000   1995.000000  
75%       12.000000    400.000000      3.000000   1996.000000  
max       16.000000    700.000000      4.000000   1996.000000

In [ ]:
# 9: Categorical columns
print("Categorical columns:")
cat_cols = df.select_dtypes(include=['object']).columns.tolist()
print(cat_cols)

# Show unique values for each
for col in cat_cols[:5]:  # First 5 categorical columns
    print(f"\n{col}: {df[col].nunique()} unique values")
    print(df[col].value_counts().head())

Categorical columns:
['Month', 'DayOfWeek', 'Make', 'AccidentArea', 'DayOfWeekClaimed', 'MonthClaimed', 'Sex', 'MaritalStatus', 'Fault', 'PolicyType', 'VehicleCategory', 'VehiclePrice', 'Days:Policy-Accident', 'Days:Policy-Claim', 'PastNumberOfClaims', 'AgeOfVehicle', 'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType', 'NumberOfSuppliments', 'AddressChange-Claim', 'NumberOfCars', 'BasePolicy', 'FraudFound']

Month: 12 unique values
Month
Jan    1411
May    1367
Mar    1360
Jun    1321
Oct    1305
Name: count, dtype: int64

DayOfWeek: 7 unique values
DayOfWeek
Monday       2616
Friday       2445
Tuesday      2300
Thursday     2173
Wednesday    2159
Name: count, dtype: int64

Make: 19 unique values
Make
Pontiac      3837
Toyota       3121
Honda        2801
Mazda        2354
Chevrolet    1681
Name: count, dtype: int64

AccidentArea: 2 unique values
AccidentArea
Urban    13822
Rural     1598
Name: count, dtype: int64

DayOfWeekClaimed: 8 unique values
DayOfWeekClaimed


In [16]:
#10: Saving observations
observations = """
# Initial Data Exploration - Observations

Date: February 2026

## Dataset Overview:
- Total rows: 15,420
- Total columns: 33
- Fraud percentage: ~6.0% (≈ 930 fraudulent vs ≈ 14,490 legitimate claims)

## Key Findings:
1. Severe class imbalance (~6% fraud rate) → must prioritize F1-score / Precision-Recall and use imbalance techniques (SMOTE, class_weight, scale_pos_weight)
2. 'Age' contains invalid 0 values (likely encoding for missing) → should convert 0 → NaN or impute with median
3. 'PolicyNumber' is just a row identifier (1 to 15420) → safe to drop
4. 'Year' has almost no variation (mostly 1994–1996) → very low signal → consider dropping
5. 'Deductible' is extremely peaked (mostly 300 or 400) → limited predictive value
6. Several categorical features have natural order and should be treated as ordinal:
   - AgeOfVehicle, AgeOfPolicyHolder, PastNumberOfClaims, VehiclePrice
7. Strong fraud-signal candidates (based on domain + similar datasets):
   - Fault == "Third Party"
   - PoliceReportFiled == "No"
   - WitnessPresent == "No"
   - AgentType == "External"
   - AddressChange-Claim with recent change
   - Very short time between accident and claim
8. No obvious large-scale missing values (but watch for "none" / empty strings)
9. Temporal mismatch between accident date and claim date → good opportunity for engineered delay features

## Next Steps:
- Clean invalid Age values
- Drop useless columns (PolicyNumber, possibly Year)
- Apply proper ordinal encoding where it makes sense
- Create time-based features (days between accident & claim, weekend/night flags, etc.)
- Deep dive into fraud vs non-fraud patterns (group-by + bar plots)
- Start feature engineering focused on high-signal indicators listed above
- Prepare for modeling with stratified split + imbalance handling
"""

print(observations)


# Initial Data Exploration - Observations

Date: February 2026

## Dataset Overview:
- Total rows: 15,420
- Total columns: 33
- Fraud percentage: ~6.0% (≈ 930 fraudulent vs ≈ 14,490 legitimate claims)

## Key Findings:
1. Severe class imbalance (~6% fraud rate) → must prioritize F1-score / Precision-Recall and use imbalance techniques (SMOTE, class_weight, scale_pos_weight)
2. 'Age' contains invalid 0 values (likely encoding for missing) → should convert 0 → NaN or impute with median
3. 'PolicyNumber' is just a row identifier (1 to 15420) → safe to drop
4. 'Year' has almost no variation (mostly 1994–1996) → very low signal → consider dropping
5. 'Deductible' is extremely peaked (mostly 300 or 400) → limited predictive value
6. Several categorical features have natural order and should be treated as ordinal:
   - AgeOfVehicle, AgeOfPolicyHolder, PastNumberOfClaims, VehiclePrice
7. Strong fraud-signal candidates (based on domain + similar datasets):
   - Fault == "Third Party"
   - Poli

# Initial Data Exploration

**Date:** February 01, 2026  
**Dataset:** Insurance Claims Fraud Detection  
**Source:** Kaggle (classic fraudulent insurance claims dataset)

## Quick Stats
- **Total Claims:** 15,420  
- **Fraud Cases:** ≈ 923–960 (exact count varies slightly by version)  
- **Fraud Rate:** ≈ **6.0%** (severely imbalanced)  
- **Features:** 33 columns

## Key Observations
- **Extreme class imbalance** (~6% fraud) → must use F1 / Precision-Recall AUC as primary metrics + imbalance handling (SMOTE, class_weight='balanced', scale_pos_weight)
- `Age` has invalid **0** values (very likely encoded missing/unknown) → needs cleaning (replace 0 → NaN or median of realistic ages)
- `PolicyNumber` is just a sequential ID (1 to 15,420) → **no predictive value** → safe to drop
- `Year` is nearly constant (mean ≈ 1994.87, mostly 1994–1996) → almost no variance → **drop candidate**
- `Deductible` is heavily peaked (mostly 300 or 400) → limited signal expected
- Several features have clear **ordinal** nature and should be encoded accordingly (not one-hot):
  - `AgeOfVehicle` ("new" → "more than 7")
  - `AgeOfPolicyHolder` (age bands)
  - `PastNumberOfClaims` ("none" → "more than 4")
  - `VehiclePrice` (price ranges)
- No massive missing values shown by `.isnull()`, but watch for semantic missings ("none", empty strings)
- Strong domain-signal features likely include:
  - `Fault` ("Third Party" often riskier)
  - `PoliceReportFiled` = "No"
  - `WitnessPresent` = "No"
  - `AgentType` = "External"
  - `AddressChange-Claim` (recent change = red flag)
  - Very short `Days:Policy-Accident` / `Days:Policy-Claim`
  - Young age + high-value sport vehicles

## High-Priority Columns for Next Analysis
- `FraudFound` (target – very imbalanced!)
- `Fault` (Policy Holder vs Third Party – strong domain signal)
- `PoliceReportFiled` ("No" often suspicious)
- `WitnessPresent` ("No" is common in fraud patterns)
- `AgentType` (External vs Internal)
- `AddressChange-Claim` (recent change = red flag)
- `Days:Policy-Accident` & `Days:Policy-Claim` (very short delays suspicious)
- `Age` (after cleaning the 0s)
- `AgeOfVehicle` & `AgeOfPolicyHolder` (ordinal encoding potential)
- `PastNumberOfClaims` (none vs many – behavioral signal)
- `AccidentArea` (Urban vs Rural – sometimes different fraud patterns)
- `VehicleCategory` & `VehiclePrice` (sport/high-value cars more risky)

**Next Steps (planned):**
- Clean Age = 0
- Drop PolicyNumber (+ Year?)
- Create ordinal encoding mapping for ordered categoricals
- Engineer time-lag features (accident → claim delay, weekend/night flags)
- Run grouped fraud rates on top-signal columns
- Move to deeper bivariate analysis & feature importance screening